sendrecv/gst-rust/src/main.rs

```diff
@@ -0,0 +1,426 @@
+extern crate glib;
+extern crate gstreamer as gst;
+extern crate gstreamer_sdp;
+extern crate gstreamer_sdp_sys;
+extern crate gstreamer_webrtc;
+extern crate rand;
+extern crate ws;
+#[macro_use]
+extern crate serde_json;
+
+use glib::translate::*;
+use gst::prelude::*;
+use gst::{BinExt, ElementExt};
+use rand::Rng;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Arc;
+
+#[derive(PartialEq, Eq, Debug)]
+enum AppState {
+    // AppStateUnknown = 0,
+    AppStateErr = 1,
+    ServerConnecting = 1000,
+    ServerConnectionError,
+    ServerConnected,
+    ServerRegistering = 2000,
+    ServerRegisteringError,
+    ServerRegistered,
+    ServerClosed,
+    PeerConnecting = 3000,
+    PeerConnectionError,
+    PeerConnected,
+    PeerCallNegotiating = 4000,
+    PeerCallStarted,
+    PeerCallStopping,
+    PeerCallStopped,
+    PeerCallError,
+    None,
+}
+
+fn check_plugins() -> bool {
+    let needed = vec![
+        "opus",
+        "vpx",
+        "nice",
+        "webrtc",
+        "dtls",
+        "srtp",
+        "rtpmanager",
+        "videotestsrc",
+        "audiotestsrc",
+    ];
+    let registry = gst::Registry::get();
+    let mut ret = true;
+    for plugin_name in needed {
+        let plugin = registry.find_plugin(&plugin_name.to_string());
+        if plugin.is_none() {
+            println!("Required gstreamer plugin '{}' not found", plugin_name);
+            ret = false;
+        }
+    }
+    return ret;
+}
+
+fn setup_call(app_state: &Arc<AtomicUsize>, out: &ws::Sender) -> AppState {
+    app_state.store(AppState::PeerConnecting as usize, Ordering::Relaxed);
+    println!("Setting up signalling server call with 1");
+    out.send("SESSION 1").unwrap();
+    return AppState::PeerConnecting;
+}
+
+fn register_with_server(app_state: &Arc<AtomicUsize>, out: &ws::Sender) -> AppState {
```

Comment:
Might be nice to move the app state into a struct for extensibility later instead of directly passing around that single integer everywhere (and maybe wrap that struct into an `Rc<Mutex<_>>`)